New roles

mtbc opened the pull request:
--exclude new roles work

This PR is a temporary placeholder to fix #5048 and capture new work while @pwalczysko is away.

mtbc committed:
move SQL scripts ready for DB patch 14

mtbc committed:
prepare SQL scripts for patch 14

also simplify end-script message-printing

mtbc committed:
add light admin privileges and session sudoers to OMERO model

mtbc committed:
generate new database schema for patch 14

mtbc committed:
add SQL for actual upgrade DDL

mtbc committed:
add initial light admin privileges to OMERO model

mtbc committed:
ease debugging for server-side use of enumerations

includes enumeration string value in toString()

mtbc committed:
extend the IAdmin API for light administrator privileges

mtbc committed:
add a light administrator privileges helper utility bean

mtbc committed:
allow using admin privs, other enums, in hash-based collections

mtbc committed:
implement new IAdmin methods for light admin privileges

mtbc committed:
graph traversal cannot ignore permissions for light administrators

mtbc committed:
have server session management preserve sudoer information

mtbc committed:
add ModifyUser privilege checking to IAdmin implementation

mtbc committed:
build Sudo privilege checking into session bean

mtbc committed:
build ModifyUser, WriteOwned checks into security system

mtbc committed:
fix remaining unit tests after constructor changes

mtbc committed:
add tests for working with light admin privileges via IAdmin

mtbc committed:
add testing for ModifyUser, Sudo, WriteOwned privileges

mtbc committed:
decrease log level for unknown privilege checks

mtbc committed:
thread admin privileges helper into graph requests helper

mtbc committed:
help for graphs requests with checking admin privileges

mtbc committed:
make set from LightAdminPrivileges.getSessionPrivileges immutable

mtbc committed:
make light admin privilege test code slightly clearer

mtbc committed:
add light admin privileges for Chgrp and Chown

mtbc committed:
check for Chgrp and Chown privilege in OMERO interceptor

mtbc committed:
check for Chgrp and Chown privilege in graphs requests

mtbc committed:
add integration testing for Chgrp and Chown admin privileges

mtbc committed:
fix Javadoc for WriteOwned privilege tests

mtbc committed:
simplify how tests add user to system group

mtbc committed:
move Chgrp, Chown privilege tests to list privileges in order

mtbc committed:
add initial testing for WriteFile light admin privilege

mtbc committed:
have model mock factory suffix file paths with usual separator

mtbc committed:
add WriteFile checking to basic security system

mtbc committed:
make variable final in test

mtbc committed:
remove unnecessary method annotation

mtbc committed:
add test for robustness of script service listing scripts

mtbc committed:
prevent creating objects in inaccessible groups

mtbc committed:
add light admin privilege checking to script service

mtbc committed:
extend integration tests to cover various WriteFile cases

mtbc committed:
extract integration test importFileset methods into superclass

mtbc committed:
add light admin WriteFile test for changing file checksum

mtbc committed:
fix error message to not mention null

mtbc committed:
internal service factory creation should follow group name

mtbc committed:
have import request accept and propagate group context

mtbc committed:
have import process propagate call context to import request

mtbc committed:
minor cleanup of light administrator integration tests

mtbc committed:
add WriteFile restriction test for FS import

mtbc committed:
fix reader name in import process for integration testing

mtbc committed:
add integration tests for the ReadSession privilege

mtbc committed:
allow light admin privilege check without caching session

mtbc committed:
build sudo-aware session load filter into basic ACL voter

mtbc committed:
fix minor typo in comment

mtbc committed:
build light admin security filter into basic security system

Rather than including in security filter holder, which chooses
among alternative filters, this light admin security filter
is added to a /collection/ of filters used by the basic security
system.

mtbc committed:
rework session querying to be sudo- and ReadSession-aware

mtbc committed:
rework light admin testing

- tidy abstract server test roles management
- tidy handling of sudo in light admin tests

mtbc committed:
add integration tests for the ReadSession privilege

mtbc committed:
wrap longest lines in integration test code (whitespace only)

mtbc committed:
switch internal service factory to groups by ID, not name

avoids risky session factory destruction

mtbc committed:
add IAdmin.getCurrentAdminPrivileges() with integration test

mtbc committed:
test WriteFile privilege enforcement via newTable

mtbc committed:
intercalating underscores as single character

mtbc committed:
have BasicACLVoter.allowChmod respect modify/write privileges

mtbc committed:
adjust Chgrp2I to take better account of Chgrp privilege

mtbc committed:
improve links from Javadoc

mtbc committed:
add IAdmin.createLightSystemUser method with test

mtbc committed:
better error checking/reporting for mistyped privileges

mtbc committed:
for chgrp,chown make screen-plate policy match other containers

mtbc committed:
to chgrp,chown permissions checks add light admin workaround

mtbc committed:
protect interceptor from being passed an unloaded event

mtbc committed:
protect basic ACL voter from being passed an unloaded event

mtbc committed:
fix mistake in test method Javadoc

mtbc committed:
the switch to proper FakeReader reduces retained client paths

mtbc committed:
correct accidentally duplicated constructor argument

mtbc committed:
use top-level OMERO binary repository in WriteFile testing

mtbc committed:
simplify setup noting all light admin privilege

mtbc committed:
have Javadoc reflect change in method

mtbc committed:
Merge remote-tracking branch 'remotes/origin/develop' into roles

Conflicts:
	components/blitz/src/omero/cmd/graphs/Chown2I.java
	components/model/resources/mappings/meta.ome.xml
	components/tools/OmeroJava/test/integration/AbstractServerTest.java

mtbc committed:
make security system aware of script repository UUID

mtbc committed:
map OriginalFile.repo from DB into OMERO object model

mtbc committed:
have integration tests use uniquely named files

Allows tests to run multiple times without running foul of
name uniqueness constraints.

mtbc committed:
fix original file creation for newly mapped "repo" property

Threads Hibernate session into Repository DAO file registration.

Conservatively continuing to set file repo via SQL for the interim.
Must therefore merge that change into Hibernate object state.

mtbc committed:
prevent HQL setting of OriginalFile.repo property

Before allowing write access to OriginalFile.repo we
must first think through the permissible conditions.

mtbc committed:
use unique object names in tests

mtbc committed:
add new WriteScriptRepo light admin privilege

mtbc committed:
add permissions tests for WriteScriptRepo privilege

largely mirroring WriteFile tests which apply to other repos

mtbc committed:
expand server permissions checking for WriteScriptRepo

mtbc committed:
split Delete* light admin privileges out  from Write* ones

mtbc committed:
adjust Basic ACL Voter to separate Write* from Delete*

mtbc committed:
adjust integration tests for Write* and Delete* privileges

mtbc committed:
reorder integration tests within class

mtbc committed:
use script repository (easily identified) in repo UUID test

mtbc committed:
more correctly identify the script repository among its kin

mtbc committed:
improve how tests find the desired kind of repository

mtbc committed:
re-permit users writing to world-writable groups

mtbc committed:
be clearer about expecting to find repositories

mtbc committed:
adjust "fs sets" test to expect subsets of IDs

can't use straight equality because report will also list
filesets that are in publicly readable groups

mtbc committed:
fix minor typo in comment

mtbc committed:
fix security system unit tests with dummy sessions

mtbc committed:
fix logic in AdminImpl user permissions test

mtbc committed:
to test script upload avoid working on another test's script

mtbc committed:
limit system type permissions check skipping to job types only

mtbc committed:
add ModifyGroup, ModifyGroupMembership light admin privileges

mtbc committed:
add tests for new user management light admin privileges

mtbc committed:
add core security system checks for user management privileges

mtbc committed:
add admin service checks for user management privileges

mtbc committed:
reorder integration tests

mtbc committed:
avoid needless object creation

mtbc committed:
make session reload robust against user, group name change

cf. trac 11693; cc: @joshmoore

mtbc committed:
prefix light admin privilege config property names

mtbc committed:
adjust admin service test to expect admin privilege prefixes

mtbc committed:
do __del__() on created OMERO clients in integration test

mtbc committed:
added WriteManagedRepo, DeleteManagedRepo admin privileges

mtbc committed:
introduce and wire bean for noting managed repo UUIDs

mtbc committed:
actually note managed repo UUIDs

mtbc committed:
wire OMERO interceptor into Repository DAO

mtbc committed:
give OMERO interceptor another chance to check file creation

Needs another go because now repo is set.
This can be reverted once repo is no longer set via SQL.

Because OriginalFile is originally created with null repo,
light admins require WriteFile regardless of repository.

mtbc committed:
basic sec. system checks WriteManagedRepo, DeleteManagedRepo

mtbc committed:
adjust integ. tests for WriteManagedRepo, DeleteManagedRepo

mtbc committed:
reorder integration tests

mtbc committed:
set creation time for newly created files

mtbc committed:
fix indent to match surrounding

mtbc committed:
use correct group context for finalizing import log file

mtbc committed:
adjust linebreak in comment

mtbc committed:
graph ops must allow perms check for some object properties

mtbc committed:
fix light admin script repo test to use correct repo

mtbc committed:
return graph ops to previous treatment of groups

partially reverts 229a7e01
also avoids loading OriginalFile instances

mtbc committed:
do not require delete permission for Chown light admins

mtbc committed:
remove unused imports from graphs commands

mtbc committed:
have session manager's reload perform fewer queries

suggested by @joshmoore

mtbc committed:
fix typo in comment

mtbc committed:
re-allow group owners to create users in their groups

mtbc committed:
copy user configuration along with rest of user properties

should probably have been in #4762 really

mtbc committed:
adapt OMERO interceptor perms adjustments to improve rwrwra

mtbc committed:
rename variable and use createExperimenterI helper method

mtbc committed:
add user management permissions tests

mtbc committed:
build privilege elevation checks into admin service

also simplify createLightSystemUser now we copy user configs

mtbc committed:
move privileges check into conditional that uses it

mtbc committed:
mark all privilege elevation tests as not being broken

mtbc committed:
tidy use of OmeroInterceptor.getAdminPrivileges

mtbc committed:
break very long method signature lines in test class

joshmoore commented on components/server/src/ome/security/basic/OmeroInterceptor.java:
```diff
@@ -91,6 +105,11 @@
+    /** thread-safe and may be removed without warning;
+     *  used by {@link omero.cmd.graphs.BaseGraphTraversalProcessor#deleteInstances(String, Collection)} */
+    public static SetMultimap<Session, Long> DELETED_GROUP_LINK_IDS =
+            Multimaps.synchronizedSetMultimap(HashMultimap.<Session, Long>create());
```
I know we're still in the "finding a solution" phase, but a global synchronized map worries me.

mtbc replied:
Me too. Could make private and hide behind a public method that does the `put`s but map would still have to be synchronized. (Could make `preFlush` synchronized instead but that felt even worse.) The architecture of having `Delete2` talk directly to `OmeroInterceptor` which then directly uses `dataSource` is horrifying anyway. It's been tricky to write the interceptor code while working around that most of its methods don't have any sense of which session/transaction they're working in. (Indeed, I'm only indexing by sessions at all here because the current transaction from `Delete2`'s point of view appears not to be the one the interceptor is notified of.)

mtbc committed:
add test for restricting root's privileges

@will-moore reported the issue

mtbc committed:
use roles table for barrier triggers incl. root restrictions

mtbc committed:
add current admin privilege table to database, with triggers

triggers to prevent privilege elevation

mtbc committed:
enable SqlAction to work with roles and curr. privileges tables

mtbc committed:
in OmeroInterceptor.preFlush populate DB current privileges

needn't specifically be preFlush but that does work
and beforeTransactionCompletion doesn't get called

mtbc committed:
populate role IDs table on server startup

mtbc committed:
periodically clean up current admin privileges table

mtbc committed:
move SQL scripts for final database release

mtbc committed:
generate SQL scripts for 5.3 DB

mtbc committed:
adjust manual upgrade scripts for final database version

mtbc committed:
quote % signs in SQL script generation

mtbc committed:
add comment describing database trigger function

mtbc committed:
throw ApiUsageException when restricting root's privileges

mtbc committed:
adjust _roles table to always have one row

mtbc committed:
adjust _current_admin_privileges table cleanup strategy

Odd deadlock / hang was being noted in PostgreSQL.
This now deletes rows by specific transaction ID only for
long-completed transactions.

mtbc commented:
Opened fresh version as #5141 that eliminates some early bothersome commits like those from #4957 so that the commit history is fairly clean.

@pwalczysko, @will-moore: Once CI is pleased with this then not sure how much of a pain it would be for you to adjust your branches to be just your commits against a new `roles` force-push. Absolutely feel free to push your latest then tell me to push a branch that is your commits atop the new `roles` for you to resume working from: if I am causing undue pain then I am happy to use `rebase -i` and whatnot to fix it from my end.

will-moore commented:
My branch doesn't include your changes. I just started a clean web branch, then I connect to a server that includes your work.
